src/process/base/ZLibBasedFASTQProcessorBaseClass.cpp

```diff
@@ -0,0 +1,143 @@
+#include "ZLibBasedFASTQProcessorBaseClass.h"
+#include <cstdio>
+#include <experimental/filesystem>
+#include <iostream>
+#include <unistd.h>
+#include <zlib.h>
+
+ZLibBasedFASTQProcessorBaseClass::ZLibBasedFASTQProcessorBaseClass(
+        shared_ptr<Source> fastq,
+        shared_ptr<Source> index,
+        const bool enableDebugging) :
+        enableDebugging(enableDebugging) {
+
+    // clang-tidy will complain, that zStream is not initialized by this constructor. This is done in another step.
+    fastqFile = move(fastq);
+    inputIndexFile = move(index);
+}
+
+ZLibBasedFASTQProcessorBaseClass::ZLibBasedFASTQProcessorBaseClass(
+        shared_ptr<Source> fastq,
+        shared_ptr<Sink> index,
+        const bool enableDebugging) :
+        enableDebugging(enableDebugging) {
+
+    // clang-tidy will complain, that zStream is not initialized by this constructor. This is done in another step.
+    fastqFile = move(fastq);
+    outputIndexFile = move(index);
+}
+
+bool ZLibBasedFASTQProcessorBaseClass::initializeZStream(int mode) {
+    zStream.zalloc = nullptr;
+    zStream.zfree = nullptr;
+    zStream.opaque = nullptr;
+    zStream.next_in = nullptr;
+    zStream.avail_in = 0;
+    zStream.avail_out = 0;
+
+    zlibResult = inflateInit2(&zStream, mode);
+    if (zlibResult != Z_OK) {
+        addErrorMessage("The zlib stream could not be initialized.");
+        return false;
+    }
+
+    return true;
+}
+
+bool ZLibBasedFASTQProcessorBaseClass::readCompressedDataFromSource() {
+    /* get some compressed data from input file */
+    int result = this->fastqFile->read(input, CHUNK_SIZE);
+
+    if (result == -1) {
+        this->addErrorMessage("There was an error during fread.");
+        return false;
+    } else
+        zStream.avail_in = (uInt) result;
+
+    if (zStream.avail_in == 0) {
+        this->addErrorMessage("There was no data available in the stream");
+        return false;
+    }
+    zStream.next_in = input;
+    return true;
+}
+
+/**
+ * If at end of block, consider adding an index entry (note that if
+ * data_type indicates an end-of-block, then all of the
+ * uncompressed data from that block has been delivered, and none
+ * of the compressed data after that block has been consumed,
+ * except for up to seven bits) -- the total_bytes_out == 0 provides an
+ * entry point after the zlib or gzip header, and assures that the
+ * index always has at least one access point; we avoid creating an
+ * access point after the offset block by checking bit 6 of data_type
+ *
+ * @param strm z_stream reference to the struct instance used for decompression
+ * @return true, if at end of block.
+ */
+bool ZLibBasedFASTQProcessorBaseClass::checkStreamForBlockEnd() {
+    // Also here, clang-tidy complains about bitwise operations on signed fields.
+    // However, the code is from Mark Adlers code and I do not want to change it.
+    return (zStream.data_type & 128) != 0 && !(zStream.data_type & 64) != 0;
```

Comment:
These bits "128" and "64" in `data_type` mean something. Do you know what? Are there no constants that express the meaning of these bits? If not, define them and use them here!